WalletWasabi/CoinJoin/Coordinator/Rounds/CoordinatorRound.cs

```diff
@@ -548,45 +549,43 @@ public async Task ExecuteNextPhaseAsync(RoundPhase expectedPhase, Money feePerIn
 									}
-								});
+								}).ConfigureAwait(false);
 						}
 					}
 					catch (Exception ex)
 					{
 						Logger.LogWarning($"Round ({RoundId}): {expectedPhase} timeout failed.");
 						Logger.LogWarning(ex);
 					}
-				});
+				}).ConfigureAwait(false); ;
 		}
 
 		private Money CalculateNewDenomination()
 		{
 			return Alices.Min(x => x.InputSum - x.NetworkFeeToPayAfterBaseDenomination);
 		}
 
-		public async Task ProgressToOutputRegistrationOrFailAsync(params Alice[] additionalAlicesToBan)
+		public async Task ProgressToOutputRegistrationOrFailAsync(params Alice[] alicesNotConfirmConnection)
 		{
-			// Only abort if less than two one alices are registered.
-			// What if an attacker registers all the time many alices, then drops out. He'll achieve only 2 alices to participate?
-			// If he registers many alices at InputRegistration
-			// AND never confirms in connection confirmation
-			// THEN connection confirmation will go with 2 alices in every round
-			// Therefore Alices that did not confirm, nor requested disconnection should be banned:
-
-			var alicesRemoved = await RemoveAlicesIfAnInputRefusedByMempoolAsync().ConfigureAwait(false); // So ban only those who confirmed participation, yet spent their inputs.
+			var responses = await GetTxOutForAllInputsAsync().ConfigureAwait(false);
+			var alicesSpent = responses.Where(x => x.resp is null).Select(x => x.alice).ToHashSet();
+			IEnumerable<OutPoint> inputsToBan = alicesSpent.SelectMany(x => x.Inputs).Select(y => y.Outpoint).Concat(alicesNotConfirmConnection.SelectMany(x => x.Inputs).Select(y => y.Outpoint)).Distinct();
 
-			// There's the question of alices who spent their inputs but confirmed connection and alices who we removed because their inputs were unconfirmed and the resulting mempool chain was too big.
-			// We should ban alices who spent their coins, but not the ones who ended up being in a large mempool chain.
-			// However note that even alices who spent their coins don't have to be banned, since the attack described here would still be probably very expensive. Anyway, to be sure let's ban them.
-			IEnumerable<Alice> alicesToBan = alicesRemoved.removedSpentAlices;
-			IEnumerable<OutPoint> inputsToBan = alicesToBan.SelectMany(x => x.Inputs).Select(y => y.Outpoint).Concat(additionalAlicesToBan.SelectMany(x => x.Inputs).Select(y => y.Outpoint)).Distinct();
+			var alicesNotConfirmConnectionIds = alicesNotConfirmConnection.Select(x => x.UniqueId);
 
 			if (inputsToBan.Any())
 			{
 				await UtxoReferee.BanUtxosAsync(1, DateTimeOffset.UtcNow, forceNoted: false, RoundId, inputsToBan.ToArray()).ConfigureAwait(false);
+
+				var alicesConnectionConfirmedAndSpentCount = alicesSpent.Select(x => x.UniqueId).Except(alicesNotConfirmConnectionIds).Distinct().Count();
+				if (alicesConnectionConfirmedAndSpentCount > 0)
+				{
+					Abort($"{alicesConnectionConfirmedAndSpentCount} Alices confirmed their connections but spent their inputs.");
```

Comment:
@nopara73 We should return after this, shouldn't we?

Reply:
Yes.